File: arrays/05_depth_first_search/test_depth_first_search.py
from unittest import TestCase, TestResult, TestLoader
from depth_first_search import Solution
from time import time
import numpy as np

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


class TestDepthFirstSearch(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test1_Path1(self):
        matrix = np.array([[9, 9, 4], [6, 6, 8], [2, 1, 1]])
        expected = 4
        result = self.solution.longest_increasing_path(matrix)

        self.assertEqual(result, expected)

    def test2_Path2(self):
        matrix = np.array([[3, 4, 5], [3, 2, 6], [2, 2, 1]])
        expected = 4
        result = self.solution.longest_increasing_path(matrix)

        self.assertEqual(result, expected)

    def test3_Path3(self):
        matrix = np.array([[1]])
        expected = 1
        result = self.solution.longest_increasing_path(matrix)

        self.assertEqual(result, expected)

    def test4_Path4(self):
        matrix = np.array([[1, 2]])
        expected = 2
        result = self.solution.longest_increasing_path(matrix)

        self.assertEqual(result, expected)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestDepthFirstSearch)
    result = CustomTestResult()
    suite.run(result)
    result.printTestTimes()

File: arrays/05_depth_first_search/test_depth_first_search.cpp
#include "depth_first_search.cpp"
#include <gtest/gtest.h>
#include <vector>
#include <chrono>
#include <iostream>
 
using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestDepthFirstSearch : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(TestDepthFirstSearch, test1_Path1) { 
    vector<vector<int>> matrix = {{9, 9, 4}, {6, 6, 8}, {2, 1, 1}};
    int expected = 4;
    int result = solution->longestIncreasingPath(matrix);

    ASSERT_EQ(expected, result);
}

TEST_F(TestDepthFirstSearch, test2_Path2) { 
    vector<vector<int>> matrix = {{3, 4, 5}, {3, 2, 6}, {2, 2, 1}};
    int expected = 4;
    int result = solution->longestIncreasingPath(matrix);

    ASSERT_EQ(expected, result);
}

TEST_F(TestDepthFirstSearch, test3_Path3) { 
    vector<vector<int>> matrix = {{1}};
    int expected = 1;
    int result = solution->longestIncreasingPath(matrix);

    ASSERT_EQ(expected, result);
}

TEST_F(TestDepthFirstSearch, test4_Path4) { 
    vector<vector<int>> matrix = {{1, 2}};
    int expected = 2;
    int result = solution->longestIncreasingPath(matrix);

    ASSERT_EQ(expected, result);
}


int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}